### PR TITLE
Feature Contribution: Sync Audio with Gossip/Quest Frame (#48)

### PR DESCRIPTION
**Pull request by MattS8**
Added an option to automatically stop/remove the audio associated with the current gossip/quest frame. This won't affect manually queued up quest dialog. 

Also fixed a slight error in tooltip for hiding frame option.

### CODE REVIEW

File: AI_VoiceOver/Options.lua [hunk 1]
```diff
@@ -184,10 +184,7 @@ local GeneralTab =
                     type = "toggle",
                     order = 7,
                     name = "Hide NPC Portrait",
-                    desc = "Talking NPC portrait will not appear when voice over audio is played.\n\n" ..
-                            Utils:ColorizeText("This might be useful when using other addons that replace the dialog experience, such as " ..
-                                Utils:ColorizeText("Immersion", NORMAL_FONT_COLOR_CODE) .. ".",
-                                GRAY_FONT_COLOR_CODE),
+                    desc = "Talking NPC portrait will not appear when voice over audio is played.\n\n",
```
Comment (xvwyh) [inline AI_VoiceOver/Options.lua:187]: Why remove this part, has is stopped being relevant to Immersion?
And if you're removing it, remove the newlines at the end too.

File: AI_VoiceOver/VoiceOver.lua [hunk 2]
```diff
@@ -210,8 +228,38 @@ function Addon:GOSSIP_SHOW()
...
+        guid = guid,
+        name = targetName,
+    }
+end
+
+function Addon:QUEST_FINISHED()
+    if Addon.db.profile.Audio.StopAudioOnDisengage and currentQuest then
+        for index, queuedSound in ipairs(self.soundQueue.sounds) do
+            if queuedSound.questID == currentQuest.questID then
+                self.soundQueue:RemoveSoundFromQueueUsingIndex(index)
```
Comment (xvwyh) [inline AI_VoiceOver/VoiceOver.lua:243]: Why create a separate function for deleting via index, when you could just call `RemoveSoundFromQueue(queuedSound)`?

File: AI_VoiceOver/VoiceOver.lua [hunk 3]
```diff
@@ -147,6 +151,13 @@ function Addon:QUEST_DETAIL()
         unitGUID = guid
     }
     self.soundQueue:AddSoundToQueue(soundData)
+
+    -- Save current quest for Dialog-frame snyc option
```
Comment (xvwyh) [inline AI_VoiceOver/VoiceOver.lua:155]: "snyc" typo, below too.

**Comment by xvwyh** [comment]
Gossip frame will always close (GOSSIP_CLOSED) when choosing a quest from the gossip menu. Basically with this option enabled it's impossible to have multiple gossips or quests queued at the same time, you will always be removing *at most* 1 sound from the queue (either being actively played right now, or queued, if there is anything playing ahead of it, which, with this option enabled, can only really happen if the player plays something from the questlog). As such, I think we can avoid storing
```lua
local currentGossipNPC = nil
local currentQuest = nil
```
entirely and we can just store the *last created* soundData. And when GOSSIP_CLOSED/QUEST_FINISHED is called - just remove that single soundData from the queue, no need to search through the queue manually either. As it will be impossible for there to be more than 1 soundData in the queue from the same NPC.

Correct me if I'm wrong, I haven't tested this.

File: AI_VoiceOver/Options.lua (hunk 1 above)
Comment (xvwyh) [reply AI_VoiceOver/Options.lua:187]: I was under the impression that it's only the animated portrait that was an issue with Immersion, because Immersion already provides its own animated portrait. I don't see why it's beneficial to hide the entire frame and thus have no way to see the queue and to pause/resume the sound apart from the minimap button.

File: AI_VoiceOver/VoiceOver.lua (hunk 2 above)
Comment (MattS8) [reply AI_VoiceOver/VoiceOver.lua:243]: The idea was that `QUEST_FINISHED` and `GOSSIP_CLOSED` will need to loop through the sound queue and find the correct corresponding sound index anyway. Passing the entire sound data would require two loops through the queue when only one is needed. 

Since the number of queued up sounds are small, it probably doesn't make much difference, however in practice it seemed like a good idea. `RemoveSoundFromQueue` can be used exactly as before.

Comment (xvwyh) [reply AI_VoiceOver/VoiceOver.lua:243]: > Passing the entire sound data would require two loops through the queue

How so?

**Comment by xvwyh** [comment]
I'm fine with this, since we can't figure out why does the event order differs sometimes.

I'm open for a conversation about option descriptions. Your original description from your previous PR was mentioning only the "talking head", but you were hiding the whole frame, which had more functionality in it than just the talking head. When I was making the actual option to hide the portrait I took your description literally and put in on the option that hides the "talking head", not the entire frame. If that wasn't your original intent, you're free to correct it, just because I left a comment doesn't mean that you can't disagree with it.

**Comment by MattS8** [comment]
I think the options change isn't really pertinent to this feature so I just took it out of the PR for now. 

Plus, after hearing how you explained it, I sort of agree with your line of thinking. In my head, the ideal "Immersion integration" would be no voice-over UI at all, and the audio synced to the state of the Immersion dialog (i.e. paging through gossip/quest chunks skips the audio to the proper spot). That's probably why I imagined the "Hide Entirely" option to more closely line up with the original reasoning for putting that message in. 

Since there isn't that level of integration with Immersion yet, it makes sense that people would still want the UI to control the queue of sounds, but not necessarily the portrait, like you said.

File: AI_VoiceOver/VoiceOver.lua (hunk 2 above)
Comment (MattS8) [reply AI_VoiceOver/VoiceOver.lua]: I know it's outdated at this point, but the reason was due to needing to iterate through the queue to match the soundData with the saved guid/event pair. Once the correct soundData was found, the `RemoveSoundFromQueue` would iterate again to actually remove the sound from the queue. So my options were to either duplicate the removal logic within my iteration, or create a new function that removes based on the index found from the first iteration. I chose to latter as I could simply refactor the existing removal function to make it work cleanly for both use cases.

However, as you pointed out, had I simply saved the entire soundData (as is done currently), I could just pass that into the original `RemoveSoundFromQueue`. 

TLDR: I thought I was being space-efficient and clever by not copying an entire table, but I ended up making it more complicated than it needed to be!

Comment (xvwyh) [reply AI_VoiceOver/VoiceOver.lua]: Gotcha.

**Comment by mrthinger** [comment]
@MattS8 Is this ready to be reviewed?

**Comment by MattS8** [comment]
@mrthinger yup!

**Review by mrthinger — APPROVED** [review]
